### PR TITLE
feat(closure): fix #727, add zone_externs.js for closure compiler

### LINKED ISSUE #727: Distribute closure externs for Zone
Since Zone.js is a "prollyfill" and since multiple angular apps might be on a single page, we don't want to bundle it together with the app.
This means closure compiler needs an externs file to avoid renaming Zone's symbols.
I currently have such an externs here:
https://github.com/alexeagle/closure-compiler-angular-bundling/blob/master/vendor/zone_externs.js
but it should be distributed with Zone.js so that users don't have to fetch it themselves.

### PR DESCRIPTION
**Pull request by JiaLiPassion**
1. add closure compiler externs file in dist/zone_externs.js
2. add test script in travis.
now we have 2 cases.

- normal case, can compile and keep the names defined in extern
- fail case, will throw error if the extern file is broken (definition error).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,7 +43,9 @@
   "bugs": {
     "url": "https://github.com/angular/zone.js/issues"
   },
-  "dependencies": {},
+  "dependencies": {
```
Comment (alexeagle) [inline package.json]: should be a devDependency right?

File: lib/zone.ts [hunk 2]
```diff
@@ -136,11 +136,11 @@
 interface Zone {
   /**
    *
-   * @returns {Zone} The parent Zone.
+   * @returns {?Zone} The parent Zone.
```
Comment (alexeagle) [inline lib/zone.ts]: what is the reasoning for updating all the inline types?

If the externs file is independent of the .ts file,
1) we'll have to maintain two different copies of the same info, and
2) nothing enforces the JSDoc annotations in the .ts file are correct, so we will probably fail to maintain them

angular/tsickle can generate externs from the TS types, so we could just remove the JSDoc annotations if we were using the extern generator.

I know you already did a lot of work here, sorry to have to ask this!

File: scripts/closure/closure_compiler.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+OPTS_FAILED=(
+  "--compilation_level=ADVANCED_OPTIMIZATIONS"
+  "--js_output_file=build/closure/closure-bundle.js"
+  "--rewrite_polyfills=false"
+
+  # Flags to simplify debugging.
+  "--formatting=PRETTY_PRINT"
+
+  "--externs=test/closure/zone_externs_failed.js"
```
Comment (alexeagle) [inline scripts/closure/closure_compiler.sh]: nit: I'd have one options block with the options that are in common (including the zone.closure.js input file)
then when you call compiler.jar below, give the options that differ between the two cases

File: scripts/closure/closure_compiler.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+closureFlags=$(mktemp)
+echo ${OPTS[*]} > $closureFlags
+
+closureFailedFlags=$(mktemp)
+echo ${OPTS_FAILED[*]} > $closureFailedFlags
+
+# compile closure test source file
+tsc -p .
+# Run the Google Closure compiler java runnable.
+java -jar node_modules/google-closure-compiler/compiler.jar --flagfile $closureFlags 'build/test/closure/zone.closure.js'
```
Comment (alexeagle) [inline scripts/closure/closure_compiler.sh]: nit: I prefer to have the flagfile be an actual file in this directory, then you don't have to use the fancy bash array construct that few people understand

File: test/closure/zone_externs_failed.js [hunk 5]
```diff
@@ -0,0 +1,445 @@
+/**
+* @license
+* Copyright Google Inc. All Rights Reserved.
+*
+* Use of this source code is governed by an MIT-style license that can be
+* found in the LICENSE file at https://angular.io/license
+*/
+
+/**
+ * @fileoverview Externs for zone.js
```
Comment (alexeagle) [inline test/closure/zone_externs_failed.js]: this file also looks like a lot of maintenance work. What is different between this and the other one? Could we get the same test coverage by just calling closure with no externs at all?

File: package.json (hunk 1 above)
Comment (JiaLiPassion) [inline package.json]: yes, you are right, I will change it.

File: lib/zone.ts (hunk 2 above)
Comment (JiaLiPassion) [inline lib/zone.ts]: Thank for the review, I am new to closure compiler, so I just check the document and add the type I can find.

So I just revert all those inline type changes from zone.ts, is that correct?

File: scripts/closure/closure_compiler.sh (hunk 3 above)
Comment (JiaLiPassion) [inline scripts/closure/closure_compiler.sh]: @alexeagle , yes, you are right, I will make it change that way.

File: test/closure/zone_externs_failed.js (hunk 5 above)
Comment (JiaLiPassion) [inline test/closure/zone_externs_failed.js]: Yeah, you are right , I will change it that way.

**Comment by JiaLiPassion** [comment]
@alexeagle , thank you the review, I will make the changes of the comments you write above, and about the dist/zone_externs.js, is that ok? It can keep the name unchanged, I am not sure the `@type @interface @record` definitions are correct or not, please help me to check it , thank you.

File: scripts/closure/closure_compiler.sh (hunk 4 above)
Comment (JiaLiPassion) [inline scripts/closure/closure_compiler.sh]: yes, I just copied from material, I also prefer use flag directly, I will modify it.

File: dist/zone_externs.js [hunk 6]
```diff
@@ -0,0 +1,442 @@
+/**
+* @license
+* Copyright Google Inc. All Rights Reserved.
+*
+* Use of this source code is governed by an MIT-style license that can be
+* found in the LICENSE file at https://angular.io/license
+*/
+
+/**
+ * @fileoverview Externs for zone.js
```
Comment (alexeagle) [inline dist/zone_externs.js]: we can verify that these work in a real project, like overwrite https://github.com/alexeagle/closure-compiler-angular-bundling/blob/master/vendor/zone_externs.js with yours and re-build

File: package.json (hunk 1 above)
Comment (alexeagle) [inline package.json]: and you can remove the dependencies{} block too

File: package.json [hunk 7]
```diff
@@ -52,6 +53,7 @@
     "concurrently": "^2.2.0",
     "conventional-changelog": "^1.1.0",
     "es6-promise": "^3.0.2",
+    "google-closure-compiler": "^20170218.0.0",
```
Comment (alexeagle) [inline package.json]: a new April release just came out, may as well upgrade this now

File: scripts/closure/closure_compiler.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+if [ $? -eq 0 ]
+then
+  echo "Successfully pass closure compiler with zone externs"
+else
+  echo "failed to pass closure compiler with zone externs" 
+  exit 1
+fi
+
+# Run the Google Closure compiler java runnable with error zone externs.
```
Comment (alexeagle) [inline scripts/closure/closure_compiler.sh]: update comment - "without externs"

File: dist/zone_externs.js (hunk 6 above)
Comment (JiaLiPassion) [inline dist/zone_externs.js]: sure, I will check it too

File: package.json (hunk 7 above)
Comment (JiaLiPassion) [inline package.json]: Ok, got it.

File: scripts/closure/closure_compiler.sh (hunk 8 above)
Comment (JiaLiPassion) [inline scripts/closure/closure_compiler.sh]: Ok, got it.

**Comment by JiaLiPassion** [comment]
@alexeagle, I have modified the comment and update package.json, please review.

File: scripts/closure/closure_compiler.sh [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+# Run the Google Closure compiler java runnable without zone externs.
+java -jar node_modules/google-closure-compiler/compiler.jar --flagfile 'scripts/closure/closure_flagfile'
+
+# Check whether the name of Zone exposed API is lost correctly or not
+node build/closure/closure-bundle.js
+
+if [ $? -eq 1 ]
+then
+  echo "Successfully detect closure compiler error with wrong zone externs"
```
Comment (alexeagle) [inline scripts/closure/closure_compiler.sh]: here too, wrong -> missing

File: scripts/closure/closure_compiler.sh [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+else
+  echo "failed to detect closure compiler error with wrong zone externs"
```
Comment (alexeagle) [inline scripts/closure/closure_compiler.sh]: and here

Comment (JiaLiPassion) [inline scripts/closure/closure_compiler.sh]: @alexeagle , thanks, I will modify it now.

**Pull request merged**